{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57fb2d0a",
   "metadata": {},
   "source": [
    "# Requêtes pour la recherche de workflows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "599e4421",
   "metadata": {},
   "outputs": [],
   "source": [
    "from py2neo import *\n",
    "import json\n",
    "IP=\"bolt://localhost:7687\"\n",
    "USER=\"neo4j\"\n",
    "PASSWORD=\"pass\"\n",
    "graph_db = Graph(IP, auth=(USER, PASSWORD))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c40e36dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": " path                                                                                                                                             \n--------------------------------------------------------------------------------------------------------------------------------------------------\n (_768)-[:inputOf {format: []}]->(_649)-[:outputOf {format: ['{\"term\":\"Stockholm format\",\"uri\":\"http://edamontology.org/format_1961\"}']}]->(_770) ",
      "text/html": "<table><tr><th>path</th></tr><tr><td style=\"text-align:left\">(_768)-[:inputOf {format: []}]-&gt;(_649)-[:outputOf {format: [&#039;{&quot;term&quot;:&quot;Stockholm format&quot;,&quot;uri&quot;:&quot;http://edamontology.org/format_1961&quot;}&#039;]}]-&gt;(_770)</td></tr></table>"
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Exemple de requête qui part de \"Sequence set\" pour aller vers un \"Alignment\"\n",
    "# On cherche les chemins passant par au plus 4 relations (*1..4), tout en triant les realtions selon leur type (inputOf, outputOf)\n",
    "# Enfin on trie les chemins pour éviter de passer 2 fois par la même fonction\n",
    "graph_db.run(\"\"\"\n",
    "        MATCH path=(i:IO {term: 'Sequence set'})-[:inputOf|outputOf *1..4]->(o:IO {term: 'Alignment'})\n",
    "        WHERE NONE (n IN nodes(path) WHERE size([x IN nodes(path) WHERE n = x]) > 1 )\n",
    "        RETURN path\n",
    "        LIMIT 10\n",
    "    \"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "b9951e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def workflow_simple(entree, sortie, longueur_max, limit):\n",
    "    request = f\"\"\"\n",
    "        MATCH path=(i:IO {{term: '{inputTerm}'}})-[:inputOf|outputOf *1..{profMax}]->(o:IO {{term: '{outputTerm}'}})\n",
    "        WHERE NONE (n IN nodes(path) WHERE size([x IN nodes(path) WHERE n = x]) > 1 )\n",
    "        RETURN path AS p ORDER BY length(path) ASC\n",
    "        LIMIT {limit}\n",
    "    \"\"\"\n",
    "    return graph_db.run(request)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "da6a8322",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'workflow_simple' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mNameError\u001B[0m                                 Traceback (most recent call last)",
      "Input \u001B[0;32mIn [14]\u001B[0m, in \u001B[0;36m<module>\u001B[0;34m\u001B[0m\n\u001B[1;32m      5\u001B[0m profMax \u001B[38;5;241m=\u001B[39m \u001B[38;5;241m5\u001B[39m\n\u001B[1;32m      6\u001B[0m limit \u001B[38;5;241m=\u001B[39m \u001B[38;5;241m100\u001B[39m\n\u001B[0;32m----> 8\u001B[0m \u001B[38;5;28mprint\u001B[39m(\u001B[43mworkflow_simple\u001B[49m(inputTerm, outputTerm, profMax, limit))\n",
      "\u001B[0;31mNameError\u001B[0m: name 'workflow_simple' is not defined"
     ]
    }
   ],
   "source": [
    "# Variante en précisant des parametres\n",
    "\n",
    "inputTerm = \"Sequence set\"\n",
    "outputTerm = \"Alignment\"\n",
    "profMax = 5\n",
    "limit = 100\n",
    "\n",
    "print(workflow_simple(inputTerm, outputTerm, profMax, limit))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0e45c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ajouter contrainte qui passe par un outil précis\n",
    "# logiciels sur un système précis\n",
    "# logiciels avec language précis\n",
    "# associés à un terme d'ontologie précis\n",
    "# mots clés pertinents associés au workflow ? naviguer dans l'arbre de l'ontologie ?\n",
    "# Calcul du score du workflow ?\n",
    "# longueur du workflow, termes en commun entre les outils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "07a68955",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Node': '{\"id\":\"0\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"description\":\"Xconnector is a software package designed to easily retrieve, and visualize metabolomics data from different database sources.\",\"toolType\":[\"Plug-in\"],\"license\":\"Other\",\"elixirCommunity\":[],\"validated\":0,\"lastUpdate\":\"2021-10-05T21:35:11.947166Z\",\"homepageStatus\":0,\"name\":\"Xconnector\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-05T21:35:11.944560Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://github.com/Proteomicslab57357/Xconnector\"}}'}, {'Node': '{\"id\":\"1\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"description\":\"xTea (comprehensive transposable element analyzer) is designed to identify TE insertions from paired-end Illumina reads, barcode linked-reads, long reads (PacBio or Nanopore), or hybrid data from different sequencing platforms and takes whole-exome sequencing (WES) or whole-genome sequencing (WGS) data as input.\",\"toolType\":[\"Plug-in\"],\"license\":\"Other\",\"validated\":0,\"elixirCommunity\":[],\"lastUpdate\":\"2021-10-05T21:16:18.572546Z\",\"homepageStatus\":0,\"name\":\"xTea\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-05T21:16:18.570152Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://github.com/parklab/xTea\"}}'}, {'Node': '{\"id\":\"2\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"description\":\"yanocomp does detection of RNA modifications from nanopore direct RNA sequencing (DRS) data which has been \\\\\"eventaligned\\\\\" using nanopolish. It uses a comparative approach with General mixture models similar to those used by nanocompore and xpore.\",\"toolType\":[\"Plug-in\"],\"license\":\"MIT\",\"validated\":0,\"elixirCommunity\":[],\"lastUpdate\":\"2021-10-05T20:56:14.552489Z\",\"homepageStatus\":0,\"name\":\"Yanocomp\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-05T20:56:14.549954Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"http://www.github.com/bartongroup/yanocomp\"}}'}, {'Node': '{\"id\":\"3\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"description\":\"A computational pipeline of integrative multi-omics analyses for predicting cell-type disease genes and regulatory networks.\",\"toolType\":[\"Plug-in\"],\"license\":\"Not licensed\",\"validated\":0,\"elixirCommunity\":[],\"lastUpdate\":\"2021-10-05T14:41:28.230094Z\",\"homepageStatus\":0,\"name\":\"scGRNom\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-05T14:41:28.227766Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://github.com/daifengwanglab/scGRNom\"}}'}, {'Node': '{\"id\":\"4\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"accessibility\":\"Open access\",\"description\":\"A rank-based ensemble learning algorithm for cell type identification of single-cell RNA sequencing in cancer that can be used across different data platforms, using a combination of gene expression rank-based method and majority vote ensemble machine-learning probability-based prediction method.\",\"toolType\":[\"Plug-in\"],\"license\":\"MIT\",\"elixirCommunity\":[],\"validated\":0,\"lastUpdate\":\"2021-10-05T14:30:24.459031Z\",\"homepageStatus\":0,\"name\":\"scDetect\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-05T14:26:08.372718Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://github.com/IVDgenomicslab/scDetect/\"}}'}, {'Node': '{\"id\":\"5\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"description\":\"SCAFE (Single Cell Analysis of Five-prime Ends) provides an end-to-end solution for processing of single cell 5’end RNA-seq data.\",\"toolType\":[\"Plug-in\"],\"license\":\"MIT\",\"elixirCommunity\":[],\"validated\":0,\"lastUpdate\":\"2021-10-05T14:09:47.116214Z\",\"homepageStatus\":0,\"name\":\"SCAFE\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-04T14:51:48.826659Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://github.com/chung-lab/scafe\"}}'}, {'Node': '{\"id\":\"6\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"description\":\"scAMACE (integrative Analysis of single-cell Methylation, chromatin ACcessibility, and gene Expression). Python implementation (both CPU and GPU version) to a model-based approach to the joint analysis of single-cell data on chromatin accessibility, gene expression and methylation.\",\"toolType\":[\"Plug-in\"],\"license\":\"MIT\",\"elixirCommunity\":[],\"validated\":0,\"lastUpdate\":\"2021-10-05T14:07:11.863951Z\",\"homepageStatus\":0,\"name\":\"scAMACE\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-04T15:20:57.396142Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://github.com/cuhklinlab/scAMACE_py\"}}'}, {'Node': '{\"id\":\"7\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"description\":\"A neural network-based tool for the automated detection of extracellular vesicles in TEM images.\\\\n\\\\nScanEV - Extracellular Vesicle Scanner.\\\\n\\\\nUpload up to 10 images for detection.\",\"toolType\":[\"Web application\"],\"license\":\"MIT\",\"elixirCommunity\":[],\"validated\":0,\"lastUpdate\":\"2021-10-05T14:06:20.557472Z\",\"homepageStatus\":0,\"name\":\"ScanEV\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-04T15:47:05.204842Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://bioeng.ru/scanev\"}}'}, {'Node': '{\"id\":\"8\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"accessibility\":\"Open access\",\"description\":\"A method for exploratory analysis of cell-cell communication based on single cell RNA sequencing data.\",\"toolType\":[\"Plug-in\"],\"license\":\"MIT\",\"elixirCommunity\":[],\"validated\":0,\"lastUpdate\":\"2021-10-05T14:04:28.279691Z\",\"homepageStatus\":0,\"name\":\"scConnect\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-04T20:31:47.771506Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://github.com/JonETJakobsson/scConnect\"}}'}, {'Node': '{\"id\":\"9\",\"type\":\"node\",\"labels\":[\"Tool\"],\"properties\":{\"owner\":\"Jennifer\",\"cost\":\"Free of charge\",\"description\":\"scDART (single cell Deep learning model for ATAC-Seq and RNA-Seq Trajectory integration) is a scalable deep learning framework that embed the two data modalities of single cells, scRNA-seq and scATAC-seq data, into a shared low-dimensional latent space while preserving cell trajectory structures. Furthermore, scDART learns a nonlinear function represented by a neural network encoding the cross-modality relationship simultaneously when learning the latent space representations of the integrated dataset.\",\"toolType\":[\"Plug-in\"],\"license\":\"GPL-3.0\",\"elixirCommunity\":[],\"validated\":0,\"lastUpdate\":\"2021-10-05T14:03:09.093816Z\",\"homepageStatus\":0,\"name\":\"scDART\",\"elixirNode\":[],\"elixirPlatform\":[],\"additionDate\":\"2021-10-04T20:49:44.664451Z\",\"collectionID\":[],\"confidenceFlag\":\"tool\",\"elixirBadge\":0,\"homepage\":\"https://github.com/PeterZZQ/scDART\"}}'}]\n"
     ]
    }
   ],
   "source": [
    "# Vérificateur de terme\n",
    "def verif(term):\n",
    "    res = graph_db.run(f\"\"\"MATCH (t) RETURN apoc.convert.toJson(t) AS Node LIMIT 10\"\"\").data()\n",
    "    return res\n",
    "\n",
    "print(verif(\"ADNet\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b263da",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Recherche des données en plsu du resultat de la requête\n",
    "# TODO aller chercher les relations également (même si elles sont bien vides pour l'instant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ed1fe598",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'NoneType' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "Input \u001B[0;32mIn [20]\u001B[0m, in \u001B[0;36m<module>\u001B[0;34m\u001B[0m\n\u001B[1;32m     45\u001B[0m res \u001B[38;5;241m=\u001B[39m requete_format()\u001B[38;5;241m.\u001B[39mevaluate()\n\u001B[1;32m     46\u001B[0m \u001B[38;5;28mprint\u001B[39m(res)\n\u001B[0;32m---> 47\u001B[0m \u001B[38;5;28mprint\u001B[39m(json\u001B[38;5;241m.\u001B[39mdumps(\u001B[43mres\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mdatas\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m[\u001B[38;5;241m0\u001B[39m], indent\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m2\u001B[39m))\n",
      "\u001B[0;31mTypeError\u001B[0m: 'NoneType' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "def requete_format():\n",
    "    return graph_db.run(\"\"\"\n",
    "// Première requête renvoyant le résultat sous forme d'une liste de chemins (juste les ID)\n",
    "CALL {\n",
    "    MATCH path=(i:IO {term: 'Sequence'})-[:inputOf|outputOf *1..5]->(o:IO {term: 'Alignment'})\n",
    "    WHERE NONE (n IN nodes(path) WHERE size([x IN nodes(path) WHERE n = x]) > 1 )\n",
    "    WITH apoc.path.elements(path) AS pathAsList\n",
    "    WITH collect([x in pathAsList | ID(x)]) AS pathAsIdList\n",
    "    RETURN pathAsIdList AS pathList\n",
    "}\n",
    "// Post traitement pour aller chercher les données en lien avec les outils proposés\n",
    "CALL {\n",
    "    WITH pathList // importation de la variable dans la sous requête\n",
    "    UNWIND pathList AS pathAsIdList\n",
    "    UNWIND pathAsIdList AS stepID // On parcourt tout les noeuds de chaque chemin\n",
    "    MATCH (i)--(t:Tool)--(f:Function) WHERE ID(f)=stepID // on va chercher le noeud et ses infos, uniquement si on est bien sur une Function (et non un IO)\n",
    "\n",
    "    // Plusieurs étapes en une ligne:\n",
    "    // - On reparcours tout les noeuds trouvés\n",
    "    // - Pour chaque on va chercher son ID et son label\n",
    "    // - Puis on les ajoute dans un format de Json correct\n",
    "    // - Enfin on met toutes ces infos dans une liste (définie par compréhension) (renommée data)\n",
    "    // - on ajoute aussi l'ID pour la fonction utilisée et l'outil\n",
    "    WITH [ n in collect(i) | { id:ID(n), label:labels(n)[0], properties: properties(n)}] AS data,\n",
    "       { id:ID(f), label:labels(f)[0], properties: properties(f) } AS function,\n",
    "       { id:ID(t), label:labels(t)[0], properties: properties(t) } AS tool\n",
    "\n",
    "    return { function: function, tool: tool, data:  data } AS nodes\n",
    "}\n",
    "WITH collect(nodes) AS nodes, pathList\n",
    "CALL {\n",
    "    WITH pathList // importation de la variable dans la sous requête\n",
    "    UNWIND pathList AS pathAsIdList\n",
    "    UNWIND pathAsIdList AS stepID // On parcourt tout les noeuds de chaque chemin\n",
    "    WITH collect( DISTINCT stepID) as filtered\n",
    "    UNWIND filtered AS stepID\n",
    "    MATCH (a)-[r]->(b) WHERE ID(r)=stepID // on va chercher la relation\n",
    "    WITH ID(a) as s, ID(b) as e, r\n",
    "    WITH { id:ID(r), start: s, end: e, properties: properties(r) } AS relationship\n",
    "    RETURN relationship\n",
    "}\n",
    "    WITH collect(relationship) AS relationships, nodes, pathList\n",
    "    RETURN { paths: pathList, nodes: nodes, relationships: relationships } AS result\n",
    "    \"\"\")\n",
    "res = requete_format().evaluate()\n",
    "print(res)\n",
    "print(json.dumps(res['datas'][0], indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "630e5bd9",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'NoneType' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "Input \u001B[0;32mIn [9]\u001B[0m, in \u001B[0;36m<module>\u001B[0;34m\u001B[0m\n\u001B[0;32m----> 1\u001B[0m datas \u001B[38;5;241m=\u001B[39m \u001B[43mres\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mnodes\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\n\u001B[1;32m      2\u001B[0m s \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mset\u001B[39m()\n\u001B[1;32m      3\u001B[0m t \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mset\u001B[39m()\n",
      "\u001B[0;31mTypeError\u001B[0m: 'NoneType' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "datas = res['nodes']\n",
    "s = set()\n",
    "t = set()\n",
    "r = set()\n",
    "cpt=0\n",
    "for d in datas:\n",
    "    if (s.__contains__(d['function']['id'])):\n",
    "        print(\"Error, duplicated datas\")\n",
    "    else:\n",
    "        s.add(d['function']['id'])\n",
    "    if (t.__contains__(d['tool']['id'])):\n",
    "        print(\"Error duplicated tool\")\n",
    "    else:\n",
    "        t.add(d['tool']['id'])\n",
    "\n",
    "print(\"Nb de fonctions : \", s.__len__())\n",
    "print(\"Nb outils différents : \", t.__len__())\n",
    "paths = res['paths']\n",
    "sf = set()\n",
    "cpt=0\n",
    "for p in paths:\n",
    "    for step in p:\n",
    "        sf.add(step)\n",
    "\n",
    "print(\"fin nb d'étapes différentes: \", sf.__len__())\n",
    "\n",
    "rels = res['relationships']\n",
    "for rel in rels:\n",
    "    r.add(rel['id'])\n",
    "\n",
    "print(\"Nb relations : \", r.__len__())\n",
    "\n",
    "print(json.dumps(res, indent=2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c255d536",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null\n"
     ]
    }
   ],
   "source": [
    "def requete_format():\n",
    "    return graph_db.run(\"\"\"\n",
    "// Première requête renvoyant le résultat sous forme d'une liste de chemins (juste les ID)\n",
    "CALL {\n",
    "    MATCH path=(i:IO {term: 'Sequence'})-[:inputOf|outputOf *1..5]->(o:IO {term: 'Alignment'})\n",
    "    WHERE NONE (n IN nodes(path) WHERE size([x IN nodes(path) WHERE n = x]) > 1 )\n",
    "    WITH apoc.path.elements(path) AS pathAsList\n",
    "    WITH collect([x in pathAsList | ID(x)]) AS pathAsIdList\n",
    "    RETURN pathAsIdList AS pathList\n",
    "}\n",
    "// Post traitement pour aller chercher les données en lien avec les outils proposés\n",
    "CALL {\n",
    "    WITH pathList // importation de la variable dans la sous requête\n",
    "    UNWIND pathList AS pathAsIdList\n",
    "    UNWIND pathAsIdList AS stepID // On parcourt tout les noeuds de chaque chemin\n",
    "    MATCH (t:Tool)--(f:Function) WHERE ID(f)=stepID // on va chercher le noeud et ses infos, uniquement si on est bien sur une Function (et non un IO)\n",
    "\n",
    "    // Plusieurs étapes en une ligne:\n",
    "    // - On reparcours tout les noeuds trouvés\n",
    "    // - Pour chaque on va chercher son ID et son label\n",
    "    // - Puis on les ajoute dans un format de Json correct\n",
    "    // - Enfin on met toutes ces infos dans une liste (définie par compréhension) (renommée data)\n",
    "    // - on ajoute aussi l'ID pour la fonction utilisée et l'outil\n",
    "    WITH { id:ID(f), label:labels(f)[0], properties: properties(f) } AS function,\n",
    "       { id:ID(t), label:labels(t)[0], properties: properties(t) } AS tool, f\n",
    "       \n",
    "\n",
    "    CALL {\n",
    "        WITH f \n",
    "        MATCH (f)-[i:inputOf|outputOf]-(io: IO) \n",
    "        WITH \n",
    "            { id:ID(io), label:labels(io)[0], properties: properties(io) } AS inOut,\n",
    "            { id:ID(i), type:type(i), properties: properties(i)} AS inputOutput\n",
    "        RETURN {inOut: inOut, inputOuput: inputOutput} AS io\n",
    "    }\n",
    "    \n",
    "    WITH collect(io) AS iolist, function, tool\n",
    "\n",
    "    return { function: function, tool: tool, io: iolist} AS nodes\n",
    "}\n",
    "    WITH collect(nodes) AS nodes, pathList\n",
    "CALL {\n",
    "    WITH pathList // importation de la variable dans la sous requête\n",
    "    UNWIND pathList AS pathAsIdList\n",
    "    UNWIND pathAsIdList AS stepID // On parcourt tout les noeuds de chaque chemin\n",
    "    WITH collect( DISTINCT stepID) as filtered\n",
    "    UNWIND filtered AS stepID\n",
    "    MATCH (a)-[r]->(b) WHERE ID(r)=stepID // on va chercher la relation\n",
    "    WITH ID(a) as s, ID(b) as e, r\n",
    "    WITH { id:ID(r), start: s, end: e, properties: properties(r) } AS relationship\n",
    "    RETURN relationship\n",
    "}\n",
    "    WITH collect(relationship) AS relationships, nodes, pathList\n",
    "    RETURN { paths: pathList, nodes: nodes, relationships: relationships } AS result\n",
    "    \"\"\")\n",
    "res = requete_format().evaluate()\n",
    "print(json.dumps(res, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "703bfe7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"toolMap\": [\n",
      "    {\n",
      "      \"161374\": {\n",
      "        \"accessibility\": \"Open access\",\n",
      "        \"link\": [],\n",
      "        \"description\": \"The tool offers a wide range of multivariate methods for the exploration and integration of biological datasets with a particular focus on variable selection.\",\n",
      "        \"language\": [\n",
      "          {\n",
      "            \"name\": \"R\"\n",
      "          }\n",
      "        ],\n",
      "        \"download\": [\n",
      "          {\n",
      "            \"type\": \"Binaries\",\n",
      "            \"url\": \"https://cran.r-project.org/web/packages/mixOmics/\"\n",
      "          }\n",
      "        ],\n",
      "        \"validated\": 1,\n",
      "        \"function\": [\n",
      "          {\n",
      "            \"output\": [\n",
      "              {\n",
      "                \"data\": {\n",
      "                  \"term\": \"Score\",\n",
      "                  \"uri\": \"http://edamontology.org/data_1772\"\n",
      "                },\n",
      "                \"format\": [\n",
      "                  {\n",
      "                    \"term\": \"DSV\",\n",
      "                    \"uri\": \"http://edamontology.org/format_3751\"\n",
      "                  }\n",
      "                ]\n",
      "              }\n",
      "            ],\n",
      "            \"input\": [\n",
      "              {\n",
      "                \"data\": {\n",
      "                  \"term\": \"Matrix\",\n",
      "                  \"uri\": \"http://edamontology.org/data_2082\"\n",
      "                },\n",
      "                \"format\": [\n",
      "                  {\n",
      "                    \"term\": \"DSV\",\n",
      "                    \"uri\": \"http://edamontology.org/format_3751\"\n",
      "                  }\n",
      "                ]\n",
      "              }\n",
      "            ],\n",
      "            \"operation\": [\n",
      "              {\n",
      "                \"term\": \"Clustering\",\n",
      "                \"uri\": \"http://edamontology.org/operation_3432\"\n",
      "              },\n",
      "              {\n",
      "                \"term\": \"Visualisation\",\n",
      "                \"uri\": \"http://edamontology.org/operation_0337\"\n",
      "              },\n",
      "              {\n",
      "                \"term\": \"Principal component analysis\",\n",
      "                \"uri\": \"http://edamontology.org/operation_3960\"\n",
      "              },\n",
      "              {\n",
      "                \"term\": \"Dimensionality reduction\",\n",
      "                \"uri\": \"http://edamontology.org/operation_3935\"\n",
      "              }\n",
      "            ]\n",
      "          }\n",
      "        ],\n",
      "        \"publication\": [\n",
      "          {\n",
      "            \"metadata\": {\n",
      "              \"date\": \"2017-11-01T00:00:00Z\",\n",
      "              \"citationCount\": 758,\n",
      "              \"journal\": \"PLoS Computational Biology\",\n",
      "              \"abstract\": \"\\u00a9 2017 Rohart et al.The advent of high throughput technologies has led to a wealth of publicly available \\u2018omics data coming from different sources, such as transcriptomics, proteomics, metabolomics. Combining such large-scale biological data sets can lead to the discovery of important biological insights, provided that relevant information can be extracted in a holistic manner. Current statistical approaches have been focusing on identifying small subsets of molecules (a \\u2018molecular signature\\u2019) to explain or predict biological conditions, but mainly for a single type of \\u2018omics. In addition, commonly used methods are univariate and consider each biological feature independently. We introduce mixOmics, an R package dedicated to the multivariate analysis of biological data sets with a specific focus on data exploration, dimension reduction and visualisation. By adopting a systems biology approach, the toolkit provides a wide range of methods that statistically integrate several data sets at once to probe relationships between heterogeneous \\u2018omics data sets. Our recent methods extend Projection to Latent Structure (PLS) models for discriminant analysis, for data integration across multiple \\u2018omics data or across independent studies, and for the identification of molecular signatures. We illustrate our latest mixOmics integrative frameworks for the multivariate analyses of \\u2018omics data available from the package.\",\n",
      "              \"title\": \"mixOmics: An R package for \\u2018omics feature selection and multiple data integration\",\n",
      "              \"authors\": [\n",
      "                \"Rohart F.\",\n",
      "                \"Gautier B.\",\n",
      "                \"Singh A.\",\n",
      "                \"Le Cao K.-A.\"\n",
      "              ]\n",
      "            },\n",
      "            \"type\": [],\n",
      "            \"doi\": \"10.1371/journal.pcbi.1005752\"\n",
      "          },\n",
      "          {\n",
      "            \"metadata\": {\n",
      "              \"date\": \"2009-01-01T00:00:00Z\",\n",
      "              \"citationCount\": 256,\n",
      "              \"journal\": \"Bioinformatics\",\n",
      "              \"abstract\": \"Motivation: With the availability of many 'omics' data, such as transcriptomics, proteomics or metabolomics, the integrative or joint analysis of multiple datasets from different technology platforms is becoming crucial to unravel the relationships between different biological functional levels. However, the development of such an analysis is a major computational and technical challenge as most approaches suffer from high data dimensionality. New methodologies need to be developed and validated. Results: integrOmics efficiently performs integrative analyses of two types of 'omics' variables that are measured on the same samples. It includes a regularized version of canonical correlation analysis to enlighten correlations between two datasets, and a sparse version of partial least squares (PLS) regression that includes simultaneous variable selection in both datasets. The usefulness of both approaches has been demonstrated previously and successfully applied in various integrative studies. \\u00a9 The Author(s) 2009. Published by Oxford University Press.\",\n",
      "              \"title\": \"IntegrOmics: An R package to unravel relationships between two omics datasets\",\n",
      "              \"authors\": [\n",
      "                \"Le Cao K.-A.\",\n",
      "                \"Gonzalez I.\",\n",
      "                \"Dejean S.\"\n",
      "              ]\n",
      "            },\n",
      "            \"type\": [\n",
      "              \"Primary\"\n",
      "            ],\n",
      "            \"doi\": \"10.1093/bioinformatics/btp515\"\n",
      "          }\n",
      "        ],\n",
      "        \"elixirNode\": [],\n",
      "        \"elixirPlatform\": [],\n",
      "        \"credit\": [\n",
      "          {\n",
      "            \"typeEntity\": \"Institute\",\n",
      "            \"name\": \"unimelb.edu.au\",\n",
      "            \"typeRole\": [\n",
      "              \"Provider\"\n",
      "            ]\n",
      "          },\n",
      "          {\n",
      "            \"typeEntity\": \"Person\",\n",
      "            \"name\": \"Australia\",\n",
      "            \"email\": \"mixomics@math.univ-toulouse.fr\",\n",
      "            \"typeRole\": [\n",
      "              \"Primary contact\"\n",
      "            ]\n",
      "          }\n",
      "        ],\n",
      "        \"editPermission\": [\n",
      "          {\n",
      "            \"type\": \"group\",\n",
      "            \"authors\": [\n",
      "              \"ELIXIR-EE\",\n",
      "              \"bayjan\"\n",
      "            ]\n",
      "          }\n",
      "        ],\n",
      "        \"owner\": \"mv.schneider@unimelb.edu.au\",\n",
      "        \"cost\": \"Free of charge\",\n",
      "        \"maturity\": \"Mature\",\n",
      "        \"documentation\": [\n",
      "          {\n",
      "            \"type\": [\n",
      "              \"General\"\n",
      "            ],\n",
      "            \"url\": \"https://qfab.org/biostatistics/mixOmics\"\n",
      "          },\n",
      "          {\n",
      "            \"type\": [\n",
      "              \"User manual\"\n",
      "            ],\n",
      "            \"url\": \"https://cran.r-project.org/web/packages/mixOmics/mixOmics.pdf\"\n",
      "          }\n",
      "        ],\n",
      "        \"toolType\": [\n",
      "          \"Library\"\n",
      "        ],\n",
      "        \"license\": \"GPL-3.0\",\n",
      "        \"elixirCommunity\": [],\n",
      "        \"neo4jId\": 161374,\n",
      "        \"lastUpdate\": \"2021-09-30T10:49:02.511785Z\",\n",
      "        \"name\": \"mixomics\",\n",
      "        \"homepageStatus\": 0,\n",
      "        \"additionDate\": \"2016-10-02T22:18:58Z\",\n",
      "        \"collectionID\": [],\n",
      "        \"elixirBadge\": 0,\n",
      "        \"homepage\": \"http://mixomics.org/\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"161393\": {\n",
      "        \"accessibility\": \"Open access\",\n",
      "        \"link\": [],\n",
      "        \"description\": \"Portable Database for Metabolism. It is a format to centralizes, in a new graph-based PADMet, all information about a metabolic network. It also provides methods to import, to update, to analyse and to export data, in a library.\",\n",
      "        \"language\": [\n",
      "          {\n",
      "            \"name\": \"Python\"\n",
      "          }\n",
      "        ],\n",
      "        \"download\": [\n",
      "          {\n",
      "            \"type\": \"Software package\",\n",
      "            \"url\": \"https://github.com/AuReMe/padmet/\"\n",
      "          }\n",
      "        ],\n",
      "        \"validated\": 1,\n",
      "        \"function\": [],\n",
      "        \"publication\": [\n",
      "          {\n",
      "            \"metadata\": {\n",
      "              \"date\": \"2018-05-01T00:00:00Z\",\n",
      "              \"citationCount\": 27,\n",
      "              \"journal\": \"PLoS Computational Biology\",\n",
      "              \"abstract\": \"\\u00a9 2018 Aite et al.Genome-scale metabolic models have become the tool of choice for the global analysis of microorganism metabolism, and their reconstruction has attained high standards of quality and reliability. Improvements in this area have been accompanied by the development of some major platforms and databases, and an explosion of individual bioinformatics methods. Consequently, many recent models result from \\u201c\\u00e0 la carte\\u201d pipelines, combining the use of platforms, individual tools and biological expertise to enhance the quality of the reconstruction. Although very useful, introducing heterogeneous tools, that hardly interact with each other, causes loss of traceability and reproducibility in the reconstruction process. This represents a real obstacle, especially when considering less studied species whose metabolic reconstruction can greatly benefit from the comparison to good quality models of related organisms. This work proposes an adaptable workspace, AuReMe, for sustainable reconstructions or improvements of genome-scale metabolic models involving personalized pipelines. At each step, relevant information related to the modifications brought to the model by a method is stored. This ensures that the process is reproducible and documented regardless of the combination of tools used. Additionally, the workspace establishes a way to browse metabolic models and their metadata through the automatic generation of ad-hoc local wikis dedicated to monitoring and facilitating the process of reconstruction. AuReMe supports exploration and semantic query based on RDF databases. We illustrate how this workspace allowed handling, in an integrated way, the metabolic reconstructions of non-model organisms such as an extremophile bacterium or eukaryote algae. Among relevant applications, the latter reconstruction led to putative evolutionary insights of a metabolic pathway.\",\n",
      "              \"title\": \"Traceability, reproducibility and wiki-exploration for \\u201c\\u00e0-la-carte\\u201d reconstructions of genome-scale metabolic models\",\n",
      "              \"authors\": [\n",
      "                \"Aite M.\",\n",
      "                \"Chevallier M.\",\n",
      "                \"Frioux C.\",\n",
      "                \"Trottier C.\",\n",
      "                \"Got J.\",\n",
      "                \"Cortes M.P.\",\n",
      "                \"Mendoza S.N.\",\n",
      "                \"Carrier G.\",\n",
      "                \"Dameron O.\",\n",
      "                \"Guillaudeux N.\",\n",
      "                \"Latorre M.\",\n",
      "                \"Loira N.\",\n",
      "                \"Markov G.V.\",\n",
      "                \"Maass A.\",\n",
      "                \"Siegel A.\"\n",
      "              ]\n",
      "            },\n",
      "            \"pmcid\": \"PMC5988327\",\n",
      "            \"type\": [\n",
      "              \"Primary\"\n",
      "            ],\n",
      "            \"pmid\": \"29791443\",\n",
      "            \"doi\": \"10.1371/journal.pcbi.1006146\"\n",
      "          },\n",
      "          {\n",
      "            \"metadata\": {\n",
      "              \"date\": \"2019-11-01T00:00:00Z\",\n",
      "              \"citationCount\": 5,\n",
      "              \"journal\": \"Antioxidants\",\n",
      "              \"abstract\": \"\\u00a9 2019 by the authors. Licensee MDPI, Basel, Switzerland.Understanding growth mechanisms in brown algae is a current scientific and economic challenge that can benefit from the modeling of their metabolic networks. The sequencing of the genomes of Saccharina japonica and Cladosiphon okamuranus has provided the necessary data for the reconstruction of Genome\\u2013Scale Metabolic Networks (GSMNs). The same in silico method deployed for the GSMN reconstruction of Ectocarpus siliculosus to investigate the metabolic capabilities of these two algae, was used. Integrating metabolic profiling data from the literature, we provided functional GSMNs composed of an average of 2230 metabolites and 3370 reactions. Based on these GSMNs and previously published work, we propose a model for the biosynthetic pathways of the main carotenoids in these two algae. We highlight, on the one hand, the reactions and enzymes that have been preserved through evolution and, on the other hand, the specificities related to brown algae. Our data further indicate that, if abscisic acid is produced by Saccharina japonica, its biosynthesis pathway seems to be different in its final steps from that described in land plants. Thus, our work illustrates the potential of GSMNs reconstructions for formalizing hypotheses that can be further tested using targeted biochemical approaches.\",\n",
      "              \"title\": \"Genome\\u2013scale metabolic networks shed light on the carotenoid biosynthesis pathway in the brown algae saccharina japonica and cladosiphon okamuranus\",\n",
      "              \"authors\": [\n",
      "                \"Negre D.\",\n",
      "                \"Aite M.\",\n",
      "                \"Belcour A.\",\n",
      "                \"Frioux C.\",\n",
      "                \"Brillet-Gueguen L.\",\n",
      "                \"Liu X.\",\n",
      "                \"Bordron P.\",\n",
      "                \"Godfroy O.\",\n",
      "                \"Lipinska A.P.\",\n",
      "                \"Leblanc C.\",\n",
      "                \"Siegel A.\",\n",
      "                \"Dittami S.M.\",\n",
      "                \"Corre E.\",\n",
      "                \"Markov G.V.\"\n",
      "              ]\n",
      "            },\n",
      "            \"pmid\": \"31744163\",\n",
      "            \"type\": [],\n",
      "            \"doi\": \"10.3390/antiox8110564\"\n",
      "          },\n",
      "          {\n",
      "            \"metadata\": {\n",
      "              \"authors\": []\n",
      "            },\n",
      "            \"type\": [],\n",
      "            \"doi\": \"10.1101/462556\"\n",
      "          }\n",
      "        ],\n",
      "        \"elixirNode\": [\n",
      "          \"France\"\n",
      "        ],\n",
      "        \"elixirPlatform\": [],\n",
      "        \"credit\": [\n",
      "          {\n",
      "            \"orcidid\": \"https://orcid.org/0000-0001-9086-1485\",\n",
      "            \"name\": \"M\\u00e9ziane Aite\",\n",
      "            \"typeRole\": [\n",
      "              \"Developer\"\n",
      "            ]\n",
      "          },\n",
      "          {\n",
      "            \"orcidid\": \"https://orcid.org/0000-0001-6542-1568\",\n",
      "            \"name\": \"Anne Siegel\",\n",
      "            \"email\": \"anne.siegel@irisa.fr\",\n",
      "            \"url\": \"http://www.irisa.fr/dyliss/anne.siegel\",\n",
      "            \"typeRole\": []\n",
      "          },\n",
      "          {\n",
      "            \"orcidid\": \"https://orcid.org/0000-0002-2310-0843\",\n",
      "            \"name\": \"Jeanne Got\",\n",
      "            \"email\": \"jeanne.got@irisa.fr\",\n",
      "            \"url\": \"http://www.irisa.fr/dyliss/jeanne.got\",\n",
      "            \"typeRole\": [\n",
      "              \"Support\"\n",
      "            ]\n",
      "          }\n",
      "        ],\n",
      "        \"editPermission\": [\n",
      "          {\n",
      "            \"type\": \"private\",\n",
      "            \"authors\": []\n",
      "          }\n",
      "        ],\n",
      "        \"owner\": \"jgot\",\n",
      "        \"cost\": \"Free of charge\",\n",
      "        \"maturity\": \"Mature\",\n",
      "        \"documentation\": [\n",
      "          {\n",
      "            \"type\": [\n",
      "              \"API documentation\"\n",
      "            ],\n",
      "            \"url\": \"https://padmet.readthedocs.io/\"\n",
      "          }\n",
      "        ],\n",
      "        \"toolType\": [\n",
      "          \"Command-line tool\"\n",
      "        ],\n",
      "        \"license\": \"GPL-3.0\",\n",
      "        \"elixirCommunity\": [],\n",
      "        \"neo4jId\": 161393,\n",
      "        \"lastUpdate\": \"2021-09-28T15:38:08.761860Z\",\n",
      "        \"name\": \"PADMet\",\n",
      "        \"homepageStatus\": 0,\n",
      "        \"additionDate\": \"2019-12-13T09:44:56Z\",\n",
      "        \"collectionID\": [],\n",
      "        \"elixirBadge\": 0,\n",
      "        \"homepage\": \"https://github.com/AuReMe/padmet\"\n",
      "      }\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "def requete_toolList(idList):\n",
    "    query = \"\"\"\n",
    "    MATCH (t:Tool) WHERE ID(t) IN $idList\n",
    "    WITH t, {} as res\n",
    "CALL {\n",
    "    WITH t\n",
    "    MATCH (t)--(f:Function)\n",
    "    CALL {\n",
    "        WITH f\n",
    "        MATCH (f)-[format :inputOf]-(io:IO)\n",
    "        WITH collect({format: [form in format.format | apoc.convert.fromJsonMap(form)], data: properties(io)}) AS input, f\n",
    "        MATCH (f)-[format :outputOf]-(io:IO)\n",
    "        WITH collect({format: [form in format.format | apoc.convert.fromJsonMap(form)], data: properties(io)}) AS output, input, f\n",
    "        MATCH (f)-[:doOperation]-(op:Operation)\n",
    "        WITH collect(properties(op)) AS operation, output, input\n",
    "        RETURN input, output, operation\n",
    "    }\n",
    "    WITH collect(f{.*, input:input, output:output, operation:operation}) AS function\n",
    "    RETURN function\n",
    "}\n",
    "CALL {\n",
    "    WITH t\n",
    "    MATCH (t)-[pubID]-(pub:Publication)\n",
    "    RETURN collect(pubID{.*, metadata: properties(pub)}) AS publication\n",
    "}\n",
    "CALL { WITH t MATCH (t)--(lang:Language) RETURN collect(properties(lang)) as language }\n",
    "CALL { WITH t MATCH (t)--(link:Link) RETURN collect(properties(link)) as link }\n",
    "CALL { WITH t MATCH (t)--(d:Download) RETURN collect(properties(d)) AS download }\n",
    "CALL { WITH t MATCH (t)--(c:Credit) RETURN collect(properties(c)) AS credit }\n",
    "CALL { WITH t MATCH (t)--(d:Documentation) RETURN collect(properties(d)) AS documentation}\n",
    "CALL { WITH t MATCH (t)--(e:EditPermission) RETURN collect(properties(e)) AS editPermission}\n",
    "\n",
    "WITH collect(t{.*, neo4jId: ID(t), function: function,\n",
    "            link:link,\n",
    "            language:language,\n",
    "            publication:publication,\n",
    "            download:download,\n",
    "            credit:credit,\n",
    "            documentation:documentation,\n",
    "            editPermission:editPermission\n",
    "            }) AS tools, res\n",
    "UNWIND tools AS tool\n",
    "    WITH apoc.map.setKey(res, apoc.convert.toString(tool['neo4jId']), tool) AS res\n",
    "RETURN collect(res) AS toolMap\n",
    "            \"\"\"\n",
    "    params = {}\n",
    "    params['idList'] = idList\n",
    "    return graph_db.run(query, parameters=params)\n",
    "\n",
    "print(json.dumps(requete_toolList([161374, 161393]).data()[0], indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}